proposals/0000-type-vs-constraint.rst

```diff
@@ -0,0 +1,284 @@
+For example, if we have ::
+
+  class Default a where
+    def :: a
+
+the Core of the program will have a definition like ::
+
+  newtype Default a = MkDefault a
+
+In turn, this newtype gives rise to an axiom (coercion), like so::
+
+  axDefault :: Default a ~R# a
+
+where ``~R#`` represents primitive representational equality. Note that
+``axDefault`` is *heterogeneous*: the kind of ``Default a`` is ``Constraint``,
+whereas the kind of ``a`` is ``Type``.
+
+GHC allows us to extract out an equality relationship between *kinds* from an
+equality relationship on *types* -- and kind equalities are always nominal. To
+wit, Core allows ::
+
+  KindCo axDefault :: Constraint ~# Type
+
+Now, suppose that you could write this::
+
+  type family F a
+  type instance F Type = Int
+  type instance F Constraint = Bool
+
+If these instances were allowed, GHC could
+produce a coercion between ``Int`` and ``Bool``, thus::
+
+  Bool  ~#  F Constraint   -- By type instance F Constraint (backwards)
+        ~#  F Type         -- By KindCo axDefault
+        ~#  Int            -- By type instance F Type
+
+That would be Very, Very Bad.  So, although ``Type`` and ``Constraint`` are built
+with different (un-equal) primitive type constructors,
+
+* **GHC's type checker treats `Type` and `Constraint` as *not apart*.**
+
+That in turn makes GHC complain that the above instances overlap, and are hence illegal.
+
+
+Proposed Change Specification
+-----------------------------
+
+We propose the following new setup, not repeating any types that remains unchanged::
+
+  -- Primitive type constructors
+  type SORT :: TypeOrConstraint -> RuntimeRep -> Type
+  type IP   :: forall (r :: RuntimeRep). Symbol -> TYPE r -> CONSTRAINT r
+
+  type (=>)  :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> TYPE r2 -> Type  -- primitive
+  type (==>) :: forall (r1 :: RuntimeRep) (r2 :: RuntimeRep).
+                CONSTRAINT r1 -> CONSTRAINT r2 -> Constraint
+
+  -- Data types
+  data TypeOrConstraint = TypeLike | ConstraintLike
+
+  -- Synonyms
+  type TYPE       = SORT TypeLike
+  type CONSTRAINT = SORT ConstraintLike
+  type Constraint = CONSTRAINT LiftedRep
+
+
+Changes to the type structure
+:::::::::::::::::::::::::::::
+
+This proposal introduces ``(=>)`` and ``(==>)`` as proper type constructors, just like
+any other. Just like ``(->)``, they have kinds and can be abstracted over.
+Unlike ``FUN``, they do not take a ``Multiplicity`` argument; implicity, it is ``Many``.
+
+The ``==>`` arrow is used in two places:
+
+* In instance heads, like ``instance Eq a => Eq (Maybe a)``
+* In quantified constraints, like ``forall x. c x => Eq x``
+
+In order to be backward compatible,
+we allow writing ``=>`` instead of ``==>`` in instance heads and in quantified constraints.
+That is, the *concrete* syntax for the type construtor ``==>`` can still use ``=>``.
+However, users may also choose to write ``==>`` (imported from ``GHC.Exts``) if they choose.
```

Comment:
What about `=>`? Can it be treated like a normal operator or does it still have special magic in the parser? There is an example later that suggests it can be used in a first-class manner, but currently something like `type T = (=>)` won't parse.

Reply:
I have clarified.